newa/__init__.py

```diff
@@ -677,10 +703,25 @@ class IssueAction:  # type: ignore[no-untyped-def]
     type: IssueType = field(converter=IssueType)
     parent_id: Optional[str] = None
     job_recipe: Optional[str] = None
+    when: Optional[str] = None
 
 
 @define
 class ErratumConfig(Serializable):
     issues: list[IssueAction] = field(  # type: ignore[var-annotated]
         factory=list, converter=lambda issues: [
             IssueAction(**issue) for issue in issues])
+
+
+#
+# Recipe variables
+#
+
+class RecipeGenVar:
```

Comment:
Here I really don't know how to approach a class which may have "arbitrary" attributes.

Reply:
I'm not sure it's necessary to have one, I think we can easily use `RecipeContext` and `RecipeEnvironment`. Then, it also seems to me `RecipeJobVar` is unnecessary too, it seems to me the code around lines 484 can be simplified.

Definitely `variables.context` and `variables.environment` are basically just copying keys/values from a dictionary to a container class, `ENVIRONMENT=combination.get('environment', {})` and `CONTEXT=combination.get('context', {})` should work perfectly fine. I think this is caused by you not being aware of one trick Jinja supports in its templates, dictionary values can be retrieved by using dotted notation:

```
foo:
  bar: baz

# these are equivalent:
{{ foo["bar"] }}
{{ foo.bar }}
```

This means `CONTEXT` and `ENVIRONMENT` can be dictionaries, and still one would be able to access their content using dots, e.g. `{{ ENVIRONMENT.USER }}` would be translated to `ENVIRONMENT["USER"]` under the hood. `ENVIRONMENT` itself does not have to be a class, and we can re-use existing dictionaries.

What I don't understand is the necessity of `JOB` variable, would it make sense to drop it & use `COMPOSE=combination['compose']` instead? We already expose its `environment` and `context` attributes as standalone variables `ENVIRONMENT` and `CONTEXT`. Will there be more to store in `RecipeJobVar`?

Reply:
> I think this is caused by you not being aware of one trick Jinja supports in its templates, dictionary values can be retrieved by using dotted notation

You are right.

> What I don't understand is the necessity of `JOB` variable
It is not necessary. I just tried to follow your pattern in https://github.com/RedHatQE/newa/pull/19/files where `JOB` could be (probably incorrectly?) understood as a common parent from where one can access available parameters.

Reply:
I see. That one was mostly an example/test case for the duration of bootstrapping the needed code, demonstrating the possibilities, i.e. as an option how we may expose some attributes as standalone variables if needed, to simplify the conditions. I wouldn't say a container like that has to exist - it may, if we have an object to expose at hand already, but I wouldn't see it as a mandatory requirement. Sorry for the confusion.

Reply:
OK. Let's get rid of `JOB` for now but in the future I would like to have something like that. Let's discuss it on a mtg.

Reply:
Updated.